### PR TITLE
feat: lora finetuning

### PR DESCRIPTION
**Pull request by rheasukthanker**
#### What does this implement/fix? Explain your changes.
Add lora finetuning script

### CODE REVIEW

**Review by gabikadlecova — COMMENTED** [review]
The changes so far are good - `lora_model` is cleaner that just whittle/lora/ and sampler selection/search space collection are OK

File: whittle/lora.py [hunk 1]
```diff
@@ -0,0 +1,1008 @@
...
+    lora_r: int = 8,
+    lora_alpha: int = 16,
+    lora_dropout: float = 0.05,
+    lora_query: bool = True,
+    lora_key: bool = True,
+    lora_value: bool = True,
+    lora_projection: bool = True,
+    lora_mlp: bool = True,
+    lora_head: bool = True,
+    lora_emb: bool = True,
```
[the remark this answers is not in the document]
Comment (rheasukthanker) [reply whittle/lora.py:269]: I would keep this the way it is to align with litgpt https://github.com/Lightning-AI/litgpt/blob/v0.5.0/litgpt/finetune/lora.py#L54, but this can be done if we want to.

File: whittle/lora.py [hunk 2]
```diff
@@ -0,0 +1,1008 @@
...
+    return resume_path
+
+
+def setup(
+    checkpoint_dir: Path,
+    out_dir: Path = Path("out/finetune/lora"),
+    precision: str | None = None,
+    quantize: Literal[
+        "bnb.nf4", "bnb.nf4-dq", "bnb.fp4", "bnb.fp4-dq", "bnb.int8-training"
+    ]
```
[the remark this answers is not in the document]
Comment (rheasukthanker) [reply whittle/lora.py]: I have removed quantization(qlora) for now as it is not tested

File: whittle/lora.py [hunk 3]
```diff
@@ -0,0 +1,1008 @@
...
+    return val_loss_largest, val_loss_medium, val_loss_smallest
+
+
+def merge_lora(
+    sampling_strategy: str,
+    importance_objective: str,
+    checkpoint_dir: Path,
+    pretrained_checkpoint_dir: Path | None = None,
+    precision: str | None = None,
+) -> None:
```
[the remark this answers is not in the document]
Comment (rheasukthanker) [reply whittle/lora.py]: moved

File: whittle/sampling/sampler_factory.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return self.search_space.cast(self.grid[-1])
+
+
+def get_sampler(sampler_type, search_space, seed, num_configs, n_trials):
+    if sampler_type == "random":
+        return RandomSampler(search_space=search_space, seed=seed)
+    elif sampler_type == "grid-params":
+        return FixParamGridSampler(
+            search_space=search_space,
+            seed=42,
```
[the remark this answers is not in the document]
Comment (rheasukthanker) [reply whittle/sampling/sampler_factory.py]: I had fixed it earlier for reproducibility but I could set 42 as default then.

File: whittle/lora.py [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+    instantiate_torch_optimizer,
+    load_checkpoint,
+    num_parameters,
+    parse_devices,
+    save_hyperparameters,
+)
+from torch.utils.data import ConcatDataset, DataLoader
+from torchmetrics import RunningMean
+
+from search.search_spaces import search_spaces
```
Comment (HannanJaved) [inline whittle/lora.py:47]: Should be `from whittle.search.search_spaces import search_spaces`

File: config_hub/finetune/README.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
+## Config files
+
+The table below lists the performances you can expect from the provided config files. Note that you can achieve lower memory consumption by lowering the micro batch size as needed. In addition, you can lower the rank (`lora_r`) in the LoRA configuration files and disable LoRA for certain layers (for example, setting `lora_projection` and other LoRA layer-specific parameters to `false`).
+For more information, see the [Dealing with out-of-memory (OOM) errors](../../tutorials/oom.md) on lowering the memory requirements.
```
Comment (aaronkl) [inline config_hub/finetune/README.md:4]: dead link, seems to be from the original LitGPT documentation

File: config_hub/finetune/README.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
+## Config files
```
Comment (aaronkl) [inline config_hub/finetune/README.md:1]: Do we want this README here? It seems copy pasted from the original LitGPT readme?

File: config_hub/finetune/llama-3.1-8b/lora.yaml [hunk 8]
```diff
@@ -0,0 +1,133 @@
+
+# The path to the base model's checkpoint directory to load for finetuning. (type: <class 'Path'>, default: checkpoints/stabilityai/stablelm-base-alpha-3b)
```
Comment (aaronkl) [inline config_hub/finetune/llama-3.1-8b/lora.yaml:2]: Is this different to the original LitGPT config? If no, could we just use the original then?

File: test/test_training_strategies.py [hunk 9]
```diff
@@ -53,9 +57,10 @@ def forward(self, x):
         x_ = self.output(x_)
         return x_
 
-    def select_sub_network(self, config):
-        self.input.set_sub_network(self.input_dim, config["num_units"])
-        self.output.set_sub_network(config["num_units"], 1)
+    def set_sub_network(self, num_units):
+        print(num_units)
```
Comment (aaronkl) [inline test/test_training_strategies.py:61]: remove print

File: whittle/data/llamamini.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
```
Comment (aaronkl) [inline whittle/data/llamamini.py:1]: remove this line

File: whittle/data/llamamini.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+import torch
+from litgpt.data import DataModule, SFTDataset, get_sft_collate_fn
+from litgpt.prompts import PromptStyle
+from litgpt.tokenizer import Tokenizer
+from torch.utils.data import DataLoader, random_split
+
+
+@dataclass
+class LLaMaMini(DataModule):
+    """LIMA data module for supervised finetuning."""
```
Comment (aaronkl) [inline whittle/data/llamamini.py:18]: typo?

File: whittle/eval/utils.py [hunk 12]
```diff
@@ -23,6 +23,39 @@ def prepare_results(results, save_filepath, print_results=True):
     save_filepath.open("w", encoding="utf-8").write(json_result)
 
 
+def get_task_metric_map(dataset):
```
Comment (aaronkl) [inline whittle/eval/utils.py:26]: Could we make this into a dict? 
`task_metrics = {"wunogrande": 'acc', ....}`

File: whittle/lora.py [hunk 13]
```diff
@@ -0,0 +1,821 @@
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
```
Comment (aaronkl) [inline whittle/lora.py:1]: remove

File: whittle/lora.py [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+from whittle.eval.utils import compute_accuracy
+from whittle.lora_model.config import LoRAConfig as Config
+from whittle.lora_model.lora_gpt import GPT
+from whittle.lora_model.merge import merge_lora
+from whittle.sampling.samplers import get_sampler
+from whittle.training_strategies.base_strategy import BaseTrainingStrategy
+from whittle.training_strategies.sandwich import SandwichStrategy
+from whittle.training_strategies.standard import StandardStrategy
+
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
```
Comment (aaronkl) [inline whittle/lora.py:59]: remove, but add a comment with a link to the original code

File: whittle/lora.py [hunk 15]
```diff
@@ -0,0 +1,821 @@
...
+        logits = model(input_ids)
+        losses[k] = chunked_cross_entropy(
+            logits[..., :-1, :], targets[..., 1:], chunk_size=0
+        )
+
+    val_loss = losses.mean()
+    return val_loss
+
+
+def plot_validation_metrics(model, val_dataloader, eval, sampler):
```
Comment (aaronkl) [inline whittle/lora.py:79]: Could we rename this function (e.g compute validation_metrics_subnets or similar) since we only provide the data for plotting but do not plot directly

File: whittle/lora.py [hunk 16]
```diff
@@ -0,0 +1,821 @@
...
+    val_loss_medium = compute_loss(model, val_dataloader, eval)
+    model.reset_super_network()
+    smallest_config = sampler.get_smallest_sub_network()
+    model.set_sub_network(**smallest_config)
+    val_loss_smallest = compute_loss(model, val_dataloader, eval)
+    model.reset_super_network()
+    return val_loss_largest, val_loss_medium, val_loss_smallest
+
+
+def plot_accuracies(model, sampler, dataset, checkpoint_dir):
```
Comment (aaronkl) [inline whittle/lora.py:95]: same comment as above

File: whittle/lora.py [hunk 17]
```diff
@@ -0,0 +1,821 @@
...
+    logger_name: Literal["wandb", "tensorboard", "csv"] = "wandb",
+    seed: int = 1337,
+    access_token: str | None = None,
+    downstream_test_iters: int = 500,
+    downstream_dataset: str = "arc_easy",
+    resume: bool | Path = True,
+    dataset: str = "alpaca",
+    sampler: SamplerArgs = SamplerArgs(),
+) -> None:
+    """Finetune a model using the LoRA method.
```
Comment (aaronkl) [inline whittle/lora.py:176]:
```suggestion
    """Finetune a model using the LoRA method using super-network training strategies.
```

File: whittle/lora.py [hunk 18]
```diff
@@ -0,0 +1,821 @@
...
+    )
+
+    sampler = get_sampler(
+        sampling_strategy,
+        search_space=search_space,
+        seed=seed_sampler,
+        num_configs=num_configs,
+        n_trials=n_trials,
+    )
+    if train_strategy == "sandwich":
```
Comment (aaronkl) [inline whittle/lora.py:300]: Could we re-use  `training_strategies_cls` which would make it easier to add new training strategies in the future?

File: whittle/lora.py [hunk 19]
```diff
@@ -0,0 +1,821 @@
...
+    fabric.seed_everything(seed)  # same seed for every process to init model (FSDP)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    checkpoint_path = checkpoint_dir / "lit_model.pth"  # type: ignore
+    with fabric.init_module(empty_init=(fabric.world_size > 1)):
+        model = GPT(config)
+        if "grid_params" in sampling_strategy:
+            print("Initializing params grid....")
```
Comment (aaronkl) [inline whittle/lora.py:373]: fabric.print?

File: whittle/lora.py [hunk 20]
```diff
@@ -0,0 +1,821 @@
...
+            scheduler.step()
+            state["step_count"] += 1
+
+        total_lengths += input_ids.numel()
+        if state["iter_num"] % train.log_interval == 0:
+            loss = (
+                running_loss.compute().item()
+            )  # expensive device-to-host synchronization
+            t1 = time.perf_counter()
+            # throughput.update(
```
Comment (aaronkl) [inline whittle/lora.py:581]: remove commented files

File: whittle/sampling/random_sampler.py [hunk 21]
```diff
@@ -1,28 +1,51 @@
...
+
+class RandomSampler(BaseSampler):
     """
     RandomSampler samples configurations from a given search space using a random state.
 
     Args:
-        config_space: The search space from which to sample.
+        search_space: The search space from which to sample.
         seed: Seed for the random number generator. Defaults to None.
+        cast_search_space: Whether to cast the search space. Defaults to True.
```
Comment (aaronkl) [inline whittle/sampling/random_sampler.py:33]: Can you be more specific here? Casting the search space into what?

File: whittle/training_strategies/base_strategy.py [hunk 22]
```diff
@@ -18,10 +17,11 @@ class BaseTrainingStrategy:
 
     def __init__(
         self,
-        sampler: RandomSampler,
+        sampler,
```
Comment (aaronkl) [inline whittle/training_strategies/base_strategy.py:20]: Type is missing

File: whittle/training_strategies/random.py [hunk 23]
```diff
@@ -26,15 +26,10 @@ def __init__(self, random_samples: int = 1, **kwargs: Any):
...
         for i in range(self.random_samples):
             config = self.sampler.sample()
-            model.select_sub_network(config)
-            y_hat = model(inputs)
-            if self.kd_loss is not None:
-                loss = self.kd_loss(y_hat, outputs, y_supernet)
-            else:
-                loss = self.loss_function(y_hat, outputs)
+            model.set_sub_network(**config)
+            loss = self.compute_loss(model, inputs, outputs)
```
Comment (aaronkl) [inline whittle/training_strategies/random.py:32]: What do we do with kd loss here?

**Review by aaronkl — CHANGES_REQUESTED** [review]

File: config_hub/finetune/llama-3.1-8b/lora.yaml (hunk 8 above)
Comment (rheasukthanker) [reply config_hub/finetune/llama-3.1-8b/lora.yaml:2]: Is is different from litgpt as we have lora_emb as a parameter too

File: config_hub/finetune/README.md (hunk 7 above)
Comment (rheasukthanker) [reply config_hub/finetune/README.md:1]: I would remove this readme then

File: whittle/training_strategies/random.py (hunk 23 above)
Comment (rheasukthanker) [reply whittle/training_strategies/random.py:32]: I have not tested lora with KD loss for this PR and this should ideally be done as a separate PR after we have all the distill losses

Comment (aaronkl) [reply whittle/training_strategies/random.py:32]: ok can we open an issue to not forget about it?

Comment (aaronkl) [reply whittle/training_strategies/random.py:32]: ok can we open an issue to not forget about it?

**Review by aaronkl — APPROVED** [review]
LGTM, could you just open an issue to test the KL loss for the different training strategies?

**Comment by HannanJaved** [comment]
> LGTM, could you just open an issue to test the KL loss for the different training strategies?

Created issue #270 for this
